vpr/src/device/rr_graph_builder.h

```diff
@@ -40,6 +38,63 @@ class RRGraphBuilder {
     t_rr_graph_storage& node_storage();
     /** @brief Return a writable object for update the fast look-up of rr_node */
     RRSpatialLookup& node_lookup();
+    /** @brief Return a writable object for rr_node_metadata */
+    MetadataStorage<int>& rr_node_metadata();
+    /** @brief Return a writable object for rr_edge_metadata */
+    MetadataStorage<std::tuple<int, int, short>>& rr_edge_metadata();
+
+    MetadataStorage<int> rr_node_metadata_data() const;
+
+    MetadataStorage<std::tuple<int, int, short>> rr_edge_metadata_data() const;
+    //rr_node_metadata_clear
+    inline void rr_node_metadata_clear() {
+        rr_node_metadata_.clear();
+    }
+    inline void rr_edge_metadata_clear() {
+        rr_edge_metadata_.clear();
+    }
+    // rr_edge_metadata_size
+    inline size_t rr_edge_metadata_size() const {
+        return rr_edge_metadata_.size();
+    }
+    inline size_t rr_node_metadata_size() const {
+        return rr_node_metadata_.size();
+    }
+
+    //rr_node_metadata_find
+    template<typename LookupKey>
+    inline typename vtr::flat_map<LookupKey, t_metadata_dict>::const_iterator rr_node_metadata_find(const LookupKey& lookup_key) const {
```

Comment:
I am reading the codes about ``rr_metadata`` at

https://github.com/verilog-to-routing/vtr-verilog-to-routing/blob/36789e456ab292038b8518bf2c835c9b9dcaaa0e/vpr/src/route/rr_metadata.h#L8-L14

It does not include templates, can you explain more details why you want to include templates?
In general, we try to avoid it because it may slow down the runtime.

Reply:
I have created API's for the functions in metadata_storage.h and there a class template has been used.
https://github.com/verilog-to-routing/vtr-verilog-to-routing/blob/36789e456ab292038b8518bf2c835c9b9dcaaa0e/vpr/src/base/metadata_storage.h#L51
Inorder to pass the values to the template I am using the function template.